x-pack/plugins/ml/public/application/util/url_state.ts

```diff
@@ -58,51 +58,74 @@ export function getUrlState(search: string): Dictionary<any> {
 //   different urlStates(e.g. `_a` / `_g`) don't overwrite each other.
 export const useUrlState = (accessor: string): UrlState => {
   const history = useHistory();
-  const { search } = useLocation();
+  const { search: locationSearch } = useLocation();
+
+  // We maintain a local state of useLocation's search.
+  // This allows us to use the callback variant of setSearch()
+  // later on so we can make sure we always act on the
+  // latest url state.
+  const [search, setSearch] = useState(locationSearch);
+
+  useEffect(() => {
+    setSearch(locationSearch);
+  }, [locationSearch]);
+
+  useEffect(() => {
+    // Only push to history if something related to the accessor of this
+    // url state instance is affected (e.g. a change in '_g' should not trigger
+    // a push in the '_a' instance).
+    if (!isEqual(getUrlState(locationSearch)[accessor], getUrlState(search)[accessor])) {
```

Comment:
I'd suggest storing the result of `getUrlState` in some state value or observable and update it based on URL changes. It will let use to avoid retrieving the value from URL string on each render.

Reply:
Maybe the naming was just not as obvious, but that PR already does that: `locationSearch` refers to what we get from React Router's `useLocation`, `search` is the state value we keep. `getUrlState()` is a pure function that just parses the search string so we can compare the namespace/accessor (`_g`/`_a`), it's not touching the URL directly.

I pushed some more commits with renames to make this more obvious:

- renamed `getUrlState` to `parseUrlState`
- renamed `locationSearch` to `locationSearchString`
- renamed `search` to `searchString`

Also added a `useMemo` at the end of the hook so we return a memoized version of the parsed url state.

Reply:
yes, this commit https://github.com/elastic/kibana/pull/70576/commits/06b0235f95f3e72b905cb2837948d00bc4fc04b3 introduced `useMemo` with a parsed object,  this is what I meant 👍